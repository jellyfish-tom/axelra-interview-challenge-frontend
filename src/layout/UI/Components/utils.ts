import { __GRAY_SCALE } from '../../Theme';

export const border = (props: any) =>
  props.theme.border || props.theme.color || __GRAY_SCALE._400;

export const color = (props: any) => props.theme.color || __GRAY_SCALE._800;

export const background = (props: any) =>
  props.theme.background || props.theme.border || __GRAY_SCALE._WHITE;
