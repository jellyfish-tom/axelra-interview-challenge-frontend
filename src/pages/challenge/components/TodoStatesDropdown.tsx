import React, { useState } from "react";
import styled from "styled-components";
import { Button } from "../../../layout/UI/Components/Button";

const StyledLi = styled.li`
  float: left;
`;

const DropDownContent = styled.div`
  display: none;
  position: absolute;
  background-color: #f9f9f9;
  min-width: 10em;
  box-shadow: 0px 0.5em 1em 0px rgba(0, 0, 0, 0.2);
  z-index: 1;
  border-bottom-right-radius: 0.3em;
  border-bottom-left-radius: 0.3em;
`;

const DropDownLi = styled(StyledLi)`
  display: inline-block;
  cursor: pointer;
  margin-right: 0.5em;

  &:hover ${DropDownContent}, &:focus ${DropDownContent} {
    display: block;
  }
`;

const Item = styled.a`
  color: black;
  padding: 0.7em 1em;
  text-decoration: none;
  display: block;
  text-align: left;
  font-size: 0.9em;
  transition: all 0.5s;

  &:hover {
    background-color: #e6e6e6;
  }
`;

export interface DropdownItem {
  value: number;
  label: string;
}

export const TodoStatesDropdown = (props: {
  onClick: (state: DropdownItem) => void;
  items: DropdownItem[];
  initialState: DropdownItem;
}) => {
  const { onClick, items, initialState } = props;
  const [activeState, setActiveState] = useState(initialState);

  const onMenuItemClick = (state: DropdownItem) => {
    setActiveState(state);
    onClick(state);
  };

  return (
    <DropDownLi tabIndex={1}>
      <Button
        onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
          e.preventDefault();
        }}
      >
        {activeState.label}
      </Button>
      <DropDownContent>
        {items.map((item) => (
          <Item key={item.value} onClick={() => onMenuItemClick(item)}>
            {item.label}
          </Item>
        ))}
      </DropDownContent>
    </DropDownLi>
  );
};
