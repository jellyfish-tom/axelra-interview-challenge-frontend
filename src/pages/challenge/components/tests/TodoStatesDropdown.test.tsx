import React, { Component } from "react";
import ReactDOM, { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { shallow, configure, mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import Adapter from "enzyme-adapter-react-16";

import { withProvider } from "../../../../tests/utils";
import { TodoStatesDropdown } from "../TodoStatesDropdown";
import {
  DropdownItem,
  POSSIBLE_TODO_DROPDOWN_STATES,
  NEUTRAL_TODO_DROPDOWN_STATE,
} from "../../../../model/Dropdown";

const mockStore = configureMockStore();

const defaultProps = {
  onClick: (state: DropdownItem) => {},
  items: POSSIBLE_TODO_DROPDOWN_STATES,
  initialState: NEUTRAL_TODO_DROPDOWN_STATE,
};

describe("TodoStatesDropdown Component", () => {
  it("matches snapshot", () => {
    expect(
      shallow(
        withProvider(<TodoStatesDropdown {...defaultProps} />, mockStore({}))
      )
    ).toMatchSnapshot();
  });

  it("should render without throwing an error", () => {
    expect(
      shallow(
        withProvider(<TodoStatesDropdown {...defaultProps} />, mockStore({}))
      ).exists()
    ).toBe(true);
  });

  it("should show dropdown on 'Pick State' button hover", () => {
    const wrapper = mount(
      withProvider(<TodoStatesDropdown {...defaultProps} />, mockStore({}))
    );

    const pickStateButton = wrapper.find("button").at(0);

    // check if we found initial state button
    expect(
      pickStateButton.html().includes(NEUTRAL_TODO_DROPDOWN_STATE.label)
    ).toBe(true);
  });
});
