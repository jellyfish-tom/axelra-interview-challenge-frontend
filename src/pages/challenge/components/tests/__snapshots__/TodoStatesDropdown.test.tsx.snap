// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoStatesDropdown Component matches snapshot 1`] = `ShallowWrapper {}`;
